### PR TITLE
Resolved issue #53

### LINKED ISSUE #53: Error Handling and Logging
- Enhance error handling mechanisms across the application. 
- Implement robust error logging to track and debug issues effectively. 
- Ensure that appropriate error responses are returned to clients for better user experience.

### PR DESCRIPTION
**Pull request by 17himanshu**
fixes #53 
I have used winston for error logging into logs/error.log and also made a global middleware which handles errors and implanted a class CustomError which helps in this process of error logging

I am attaching image of log file 
image

### CODE REVIEW

**Comment by ravinder-chadha** [comment]
How will the error return to the user? can you share a screenshot of that?

**Comment by 17himanshu** [comment]
Sir I am attaching below screenshot of postman . The user or client application would receive this JSON in the response body.  We can parse it on frontend and display as per our need
image

File: backend/controllers/fileController.js
Comment (ravinder-chadha) [inline backend/controllers/fileController.js:1]: can you import logger here, and while raising errors also log them?

File: backend/controllers/userController.js
Comment (ravinder-chadha) [inline backend/controllers/userController.js:1]: Simillarly, import logger and add logging to this file as well

File: backend/controllers/fileController.js
Comment (17himanshu) [reply backend/controllers/fileController.js:1]: Sir I have this confusion in mind , like everytime there's an error I throw CutomError It is the caught by errorMiddleware which is global error handling middleware and which eventually will log the error into the log file. so doing logger.error('xyz')  will be redundant

File: backend/controllers/fileController.js [hunk 1]
```diff
@@ -2,6 +2,7 @@ const fs = require("fs");
 const File = require("../models/fileModel.js");
 const dotenv = require("dotenv");
 const path = require("path");
+const CustomError = require("../utils/CustomError.js");
```
Comment (17himanshu) [inline backend/controllers/fileController.js:5]: Sir I have imported CustomError class in usercontroller

File: backend/controllers/userController.js [hunk 2]
```diff
@@ -1,13 +1,14 @@
 const bcrypt = require('bcryptjs');
 const jwt = require('jsonwebtoken');
 const User = require('../models/userModel');
+const CustomError = require("../utils/CustomError.js");
```
Comment (17himanshu) [inline backend/controllers/userController.js:4]: similarly imported customError class in usercontroller

**Comment by 17himanshu** [comment]
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 1045.0575442843965 658.6666870117188" width="1045.0575442843965" height="658.6666870117188">
  
  
  <defs>
    <style class="style-fonts">
      @font-face {
        font-family: "Virgil";
        src: url("https://excalidraw.com/Virgil.woff2");
      }
      @font-face {
        font-family: "Cascadia";
        src: url("https://excalidraw.com/Cascadia.woff2");
      }
      @font-face {
        font-family: "Assistant";
        src: url("https://excalidraw.com/Assistant-Regular.woff2");
      }
    </style>
    
  </defs>
  <rect x="0" y="0" width="1045.0575442843965" height="658.6666870117188" fill="#ffffff"></rect><g stroke-linecap="round" transform="translate(272.2222544352212 10) rotate(0 123.33334350585938 148.66668701171875)"><path d="M32 0 C73.22 2, 109.16 0.77, 214.67 0 M32 0 C92.52 -0.89, 152.6 -0.66, 214.67 0 M214.67 0 C237.56 -0.2, 247.98 12.45, 246.67 32 M214.67 0 C235.24 -0.36, 244.89 12.14, 246.67 32 M246.67 32 C246.77 102.83, 246.1 173.84, 246.67 265.33 M246.67 32 C248.67 109.93, 247.36 188.96, 246.67 265.33 M246.67 265.33 C245.7 286.16, 237.3 297.82, 214.67 297.33 M246.67 265.33 C247.7 287.2, 234.85 298.63, 214.67 297.33 M214.67 297.33 C162.23 296.22, 110.8 296.35, 32 297.33 M214.67 297.33 C171.76 295.73, 127.72 296.68, 32 297.33 M32 297.33 C8.75 295.6, 0.36 285.89, 0 265.33 M32 297.33 C12.58 298.52, 1.88 288.08, 0 265.33 M0 265.33 C0.92 202.86, 1.81 140.29, 0 32 M0 265.33 C1.84 202.59, 1.74 140.55, 0 32 M0 32 C0.04 9.53, 10.22 -1.16, 32 0 M0 32 C-0.28 11.79, 9.43 0.6, 32 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(333.5555979410806 55.333343505859375) rotate(0 35.0899658203125 12.5)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">index.js</text></g><g transform="translate(277.8889109293618 261.6666564941406) rotate(0 119.1498794555664 12.5)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">app.use(errorMiddleware)</text></g><g stroke-linecap="round" transform="translate(689.5555979410806 13.333343505859375) rotate(0 138.66668701171875 147.33334350585938)"><path d="M32 0 C103.23 -2.54, 172.72 -2.68, 245.33 0 M32 0 C101.33 1.37, 169.94 1.71, 245.33 0 M245.33 0 C265.86 0.87, 279.08 8.78, 277.33 32 M245.33 0 C267.42 -0.92, 277.8 12.81, 277.33 32 M277.33 32 C278.27 87.94, 276.74 143.22, 277.33 262.67 M277.33 32 C277.93 94.17, 278.53 155.88, 277.33 262.67 M277.33 262.67 C277.71 284.22, 268.36 296.4, 245.33 294.67 M277.33 262.67 C277.31 284.52, 267.05 296.78, 245.33 294.67 M245.33 294.67 C165.8 296.98, 83.88 295.88, 32 294.67 M245.33 294.67 C183.31 295.53, 120.53 296.06, 32 294.67 M32 294.67 C9.3 295.49, 1.06 285.33, 0 262.67 M32 294.67 C11.19 292.92, -1.57 283.21, 0 262.67 M0 262.67 C-2.36 197.53, 0 135.79, 0 32 M0 262.67 C-1.65 173.88, -0.66 85.57, 0 32 M0 32 C1.91 10.35, 10.44 -0.24, 32 0 M0 32 C-0.78 9.08, 9.81 -1.45, 32 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(734.5555979410806 43.333343505859375) rotate(0 87.21991729736328 12.5)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">errorMiddleware.js</text></g><g transform="translate(722.8889109293618 117.33334350585938) rotate(0 105.6399154663086 25)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">global error catching </text><text x="0" y="42.519999999999996" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">middleware </text></g><g transform="translate(725.5555979410806 203.33334350585938) rotate(0 96.62991333007812 25)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">logs the error into </text><text x="0" y="42.519999999999996" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">logs/error.log</text></g><g stroke-linecap="round" transform="translate(288.2222849527993 373.3333740234375) rotate(0 142.3333282470703 137.66665649414062)"><path d="M32 0 C85.72 -3.16, 139.9 -2.55, 252.67 0 M32 0 C103.95 1.73, 176.26 1.61, 252.67 0 M252.67 0 C274.44 1.77, 286.06 8.74, 284.67 32 M252.67 0 C271.76 -1.65, 285.57 11.85, 284.67 32 M284.67 32 C286.41 99.42, 286.5 168.62, 284.67 243.33 M284.67 32 C284.77 83.26, 285.14 135.24, 284.67 243.33 M284.67 243.33 C285.54 265.27, 275.58 276.67, 252.67 275.33 M284.67 243.33 C283.97 262.54, 273.65 275.49, 252.67 275.33 M252.67 275.33 C180.08 274.45, 104.07 272.41, 32 275.33 M252.67 275.33 C189.57 273.85, 127.47 273.61, 32 275.33 M32 275.33 C11.78 274.1, -0.65 266.33, 0 243.33 M32 275.33 C12.52 273.56, 0.43 264.05, 0 243.33 M0 243.33 C0.03 166.99, 0.93 88.24, 0 32 M0 243.33 C1.78 199.93, 1.66 154.73, 0 32 M0 32 C-0.26 10.09, 11.3 -1.32, 32 0 M0 32 C-0.77 10.95, 10.16 1.71, 32 0" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(293.5555979410806 406) rotate(0 143.58987426757812 25)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">CustomError (which basically </text><text x="0" y="42.519999999999996" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">extends Error class)</text></g><g transform="translate(299.5555979410806 503.6666564941406) rotate(0 133.89988708496094 25)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">we can easily throw error </text><text x="0" y="42.519999999999996" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">rather than writing complex</text></g><g transform="translate(312.5555979410806 567.3333435058594) rotate(0 104.4098892211914 12.5)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">logger.error everytime</text></g><g stroke-linecap="round"><g transform="translate(567.9734904332278 372.3333740234375) rotate(0 60.291053753926406 -66.3520542902022)"><path d="M-0.6 -1.09 C19.62 -23.09, 99.99 -111.31, 120.39 -133.14 M1.29 0.95 C21.41 -20.78, 99.28 -109.32, 119.33 -131.83" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(567.9734904332278 372.3333740234375) rotate(0 60.291053753926406 -66.3520542902022)"><path d="M110.16 -108.57 C110.61 -114.02, 113.74 -116.88, 119.33 -131.83 M110.16 -108.57 C112.31 -116.24, 115.67 -124.18, 119.33 -131.83" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(567.9734904332278 372.3333740234375) rotate(0 60.291053753926406 -66.3520542902022)"><path d="M97.36 -119.91 C100.54 -122.77, 106.51 -123.11, 119.33 -131.83 M97.36 -119.91 C104.12 -123.51, 111.99 -127.46, 119.33 -131.83" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g transform="translate(697.7778201633029 389.0370449490015) rotate(0 168.63986206054688 125)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">so whenever an error occurs</text><text x="0" y="42.519999999999996" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">we throw new CustomError</text><text x="0" y="67.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic"></text><text x="0" y="92.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">it is catched by global error</text><text x="0" y="117.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">handling middleware</text><text x="0" y="142.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic"></text><text x="0" y="167.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">errorMiddleware under the hood </text><text x="0" y="192.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">uses winston to log the error into</text><text x="0" y="217.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">a log file</text><text x="0" y="242.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic"></text></g><g stroke-linecap="round"><g transform="translate(211.11115349663615 590.51857164171) rotate(0 38.8888888888889 -88.14815945095484)"><path d="M-0.52 -0.67 C12.36 -30.04, 64.24 -147.93, 77.26 -177.11 M1.4 1.59 C14.09 -27.57, 63.49 -146.1, 76.05 -176" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(211.11115349663615 590.51857164171) rotate(0 38.8888888888889 -88.14815945095484)"><path d="M74.87 -151.03 C74.28 -159.24, 74.39 -166.01, 76.05 -176 M74.87 -151.03 C75.71 -159.22, 76.7 -168.45, 76.05 -176" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(211.11115349663615 590.51857164171) rotate(0 38.8888888888889 -88.14815945095484)"><path d="M59.09 -157.63 C63.11 -163.83, 67.86 -168.65, 76.05 -176 M59.09 -157.63 C65.33 -163.46, 71.76 -170.4, 76.05 -176" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g transform="translate(110.7407887776692 593.4814893934463) rotate(0 75.15750122070312 26.11114501953125)"><text x="0" y="18.2986904296875" font-family="Virgil, Segoe UI Emoji" font-size="20.888916015625px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">error occurs </text><text x="0" y="44.40983544921875" font-family="Virgil, Segoe UI Emoji" font-size="20.888916015625px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">we throw error</text></g><g stroke-linecap="round"><g transform="translate(155.5555979410807 278.6666802300347) rotate(0 61.11111958821613 -2.2222137451171875)"><path d="M-1.08 -0.99 C19.35 -1.91, 101.51 -4.44, 121.97 -4.94 M0.56 1.1 C20.85 0.3, 100.87 -2.46, 120.9 -3.67" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(155.5555979410807 278.6666802300347) rotate(0 61.11111958821613 -2.2222137451171875)"><path d="M97.82 5.94 C101.96 3.3, 108.63 0.46, 120.9 -3.67 M97.82 5.94 C104.65 2.05, 114.12 -0.67, 120.9 -3.67" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g><g transform="translate(155.5555979410807 278.6666802300347) rotate(0 61.11111958821613 -2.2222137451171875)"><path d="M97.04 -11.14 C101.46 -9.45, 108.33 -7.96, 120.9 -3.67 M97.04 -11.14 C104.22 -9.28, 113.95 -6.27, 120.9 -3.67" stroke="#1e1e1e" stroke-width="2" fill="none"></path></g></g><mask></mask><g transform="translate(10 250.88881768120655) rotate(0 119.79988861083984 25)"><text x="0" y="17.52" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic">errorMiddleware mounted</text><text x="0" y="42.519999999999996" font-family="Virgil, Segoe UI Emoji" font-size="20px" fill="#1e1e1e" text-anchor="start" style="white-space: pre;" direction="ltr" dominant-baseline="alphabetic"> at bottom</text></g></svg>

**Comment by 17himanshu** [comment]
image

**Review by ravinder-chadha — APPROVED** [review]

**Pull request merged**